### PR TITLE
Implement defences against model extraction attacks

### LINKED ISSUE #233: Implement defences against model extraction attacks
Implement defences against model extraction attacks

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request introduces defences against extraction attacks presented in the literature including reverse sigmoid, random noise, class labels, high confidence, and rounding. These counter measures cannot guarantee protection from model extraction, but could make extraction slightly more difficult.

Fixes #233 

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x]  New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by TS-Lee** [comment]
Going through changes...

File: art/wrappers/output_add_random_noise.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        Make a prediction with the wrapped classifier, add random Gaussian noise and re-normalize predicted output.
+
+        :param x: Input data
+        :type x: `np.ndarray`
+        :param batch_size: Batch size
+        :type batch_size: `int`
+        :return: Rounded predictions of shape `(nb_inputs, nb_classes)`.
+        :rtype: `np.ndarray`
+        """
+        predictions = self.classifier.predict(x, batch_size=batch_size)
```
Comment (TS-Lee) [inline art/wrappers/output_add_random_noise.py]: I just wonder if we would want to pass kwargs for future change (although it's not used by classifier.predict).

Comment (beat-buesser) [reply art/wrappers/output_add_random_noise.py]: Good idea, I'll add kwargs.

**Comment by beat-buesser** [comment]
I'm working on adding support for binary classifiers.

**Comment by beat-buesser** [comment]
@TS-Lee Thank you for your review!

**Pull request merged**